### DISCUSSION TITLE
Sending Firebase Cloud Messaging messages to end-user device is not working

### DISCUSSION THREAD

**Discussion by asdeol** [kreait/firebase-php#540] [Q&A] [answered]
I am using Firebase Admin SDK for PHP to send a notification to mobile app. It was working fine before and suddenly I started getting  following error:

Message: Unable to connect to the API: cURL error 7: (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://oauth2.googleapis.com/token
File: /var/www/deolbox.com/html/covsafe/webservice/libs/vendor/kreait/firebase-php/src/Firebase/Exception/MessagingApiExceptionConverter.php
Line: 54

Trace is :
#0 /vendor/kreait/firebase-php/src/Firebase/Messaging/ApiClient.php(47): Kreait\Firebase\Exception\MessagingApiExceptionConverter->convertException(Object(GuzzleHttp\Exception\ConnectException)) 
#1 /vendor/kreait/firebase-php/src/Firebase/Messaging.php(65): Kreait\Firebase\Messaging\ApiClient->send(Object(Kreait\Firebase\Messaging\Http\Request\SendMessage)) 
#2 /include/UserOperations.php(979): Kreait\Firebase\Messaging->send(Object(Kreait\Firebase\Messaging\CloudMessage))

Looking into the json file, the URL https://oauth2.googleapis.com/token is the "token_uri".

**Comment by jeromegamez** [comment, marked as answer]
If it worked before and suddenly stopped working, it's most likely not a problem with the SDK or with the code in general, but something in the environment. As you've seen in the error, it says "unable to connect to the API", which is a networking issue. This can have several reasons:

* The remote server is down: this is something that should usually be resolved after a short time. 
* The machine (more specifically: the IP address of your machine/network) from where you are trying to make the connection is blocked by the remote server: you can check this by trying a connection from another machine in another network, e.g. from another remote server that you have access to.
* Your local network configuration is preventing you from making a connection. This could be a firewall on your computer or a Firewall in your local network.
* The IP address of the remote server could not be resolved (DNS issue)

On your machine, you can do this for debugging:

```sh
$ curl -I https://oauth2.googleapis.com/token
```

* If this returns an HTTP response between 200 and 499, the remote host is up. 
* If the HTTP response is >500, then the remote server has an error
* If you get a "Could not resolve host", then it is a DNS issue

If it works with `curl` in your terminal, but not in your PHP application, compare the cURL versions from the outputs of

```sh
$ curl --version
$ php -i | grep -i curl
```

If the shown versions don't match, especially when the PHP cURL version is lower than the system's cURL version, update the cURL extension of you PHP installation.

**Reply by asdeol** [reply]
Thanks for reply. It was firewall blocking the connection. Its working now.
